src/features/metadata-based-view/MetadataQueryAPIHelper.js

```diff
@@ -0,0 +1,140 @@
+
+import { METADATA_FIELD_TYPE_ENUM, METADATA_FIELD_TYPE_MULTISELECT, ITEM_TYPE_FILE } from '../../common/constants';
+
+import type {
+    MetadataQuery as MetadataQueryType,
+    MetadataQueryResponseData,
+    MetadataQueryResponseEntry,
+    MetadataQueryResponseEntryMetadata,
+    MetadataTemplateSchemaResponse,
+} from '../../common/types/metadataQueries';
+
+type SuccessCallback = (metadataQueryCollection: Collection) => void;
+type ErrorCallback = (e: ElementsXhrError) => void;
+
+const SELECT_TYPES: Array<typeof METADATA_FIELD_TYPE_ENUM | typeof METADATA_FIELD_TYPE_MULTISELECT> = [
+    METADATA_FIELD_TYPE_ENUM,
+    METADATA_FIELD_TYPE_MULTISELECT,
+];
+
+export default class MetadataQueryAPIHelper {
+    api: API;
+
+    metadataQueryResponseData: MetadataQueryResponseData;
+
+    metadataTemplate: ?MetadataTemplate;
+
+    templateKey: string;
+
+    templateScope: string;
+
+    constructor(api: API) {
+        this.api = api;
+    }
+
+    flattenMetadata = (metadata: MetadataQueryResponseEntryMetadata): MetadataType => {
+        const templateFields = getProp(this.metadataTemplate, 'fields', []);
+        const instance = getProp(metadata, `${this.templateScope}.${this.templateKey}`);
+
+        if (!instance) {
+            return {};
+        }
+
+        const fields = Object.keys(instance)
+            .filter(key => !key.startsWith('$'))
+            .map(key => {
+                const templateField = find(templateFields, ['key', key]);
+                const type = getProp(templateField, 'type'); // get data type
+                const field: MetadataQueryInstanceTypeField = {
+                    name: key,
+                    value: instance[key],
+                    type,
+                };
+
+                if (includes(SELECT_TYPES, type)) {
+                    // get "options" for enums or multiselects
+                    field.options = getProp(templateField, 'options');
+                }
+
+                return field;
+            });
+
+        return {
+            enterprise: {
+                fields,
+                id: instance.$id,
+            },
+        };
+    };
+
+    flattenResponseEntry = ({ item, metadata }: MetadataQueryResponseEntry): BoxItem => {
+        const { id, name, size } = item;
+
+        return {
+            id,
+            metadata: this.flattenMetadata(metadata),
+            name,
+            size,
+        };
+    };
+
+    getFlattenedDataWithTypes = (templateSchemaResponse?: MetadataTemplateSchemaResponse): Collection => {
+        this.metadataTemplate = getProp(templateSchemaResponse, 'data');
+        const { entries, next_marker }: MetadataQueryResponseData = this.metadataQueryResponseData;
+        return {
+            items: entries.map<BoxItem>(this.flattenResponseEntry),
+            nextMarker: next_marker,
+        };
+    };
+
+    filterMetdataQueryResponse = (response: MetadataQueryResponseData): MetadataQueryResponseData => {
```

Comment:
Minor, but consider to alphabetizing the method names to group similar operation types together.